geo: 3D table introspection query crashes with "invalid memory address or nil pointer dereference"

**Describe the problem**

The following query crashes:

**To Reproduce**

```sql

CREATE TABLE inspectapp_fields3d (
    id INT8 NOT NULL DEFAULT unique_rowid(),
    "point" GEOMETRY(POINTZ,4326) NOT NULL,
    pointg GEOGRAPHY(POINTZ,4326) NOT NULL,
    line GEOMETRY(LINESTRINGZ,4326) NOT NULL,
    poly GEOMETRY(POLYGONZ,4326) NOT NULL,
    CONSTRAINT "primary" PRIMARY KEY (id ASC),
    INVERTED INDEX inspectapp_fields3d_point_8e4c3179_id ("point"),
    INVERTED INDEX inspectapp_fields3d_pointg_2202a946_id (pointg),
    INVERTED INDEX inspectapp_fields3d_line_ef25c07a_id (line),
    INVERTED INDEX inspectapp_fields3d_poly_97e9595b_id (poly),
    FAMILY "primary" (id, "point", pointg, line, poly)
)

SELECT t.coord_dimension, t.srid, t.type FROM (
    SELECT * FROM geometry_columns
    UNION ALL
    SELECT * FROM geography_columns
) AS t WHERE t.f_table_name = 'inspectapp_fields3d' AND t.f_geometry_column = 'point'
```
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x35d536d]

goroutine 23125 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000885700, 0x8c90688, 0xc005227be8)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x126
panic(0x4ad8ee0, 0xb252570)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).validateRow(0xc001718f00, 0xc004d54c40, 0x7, 0x7, 0xc005975400, 0x7, 0x8, 0x70, 0x8b521a2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:471 +0xad
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).getPlanInfo.func1.1.1(0xc004d54c40, 0x7, 0x7, 0x6, 0xc003f4f900)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:534 +0x79
github.com/cockroachdb/cockroach/pkg/sql.postgisColumnsTablePopulator.func1.1(0x8dc4ef8, 0xc004fabd40, 0x523d5f8, 0x6, 0x8ddbe08, 0xc003f4f900, 0x0, 0x7f806d33f7a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pg_extension.go:86 +0x5d4
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDesc.func1(0x8dc4ef8, 0xc004fabd40, 0x523d5f8, 0x6, 0x8ddbe08, 0xc003f4f900, 0xc004d53360, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2241 +0x62
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookupInternalFromDescriptors(0x8c90688, 0xc005227868, 0xc000cb4bd0, 0x8dc4ef8, 0xc00159e360, 0x2, 0x0, 0xc0043e1c00, 0x6d, 0x6d, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2484 +0x5c2
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookupInternal(0x8c90688, 0xc005227868, 0xc000cb4bd0, 0x8dc4ef8, 0xc00159e360, 0x2, 0xc001966d00, 0xc001ddfcf8, 0xc0058fe800, 0x40)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2357 +0x110
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDescWithTableLookup(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2307
github.com/cockroachdb/cockroach/pkg/sql.forEachTableDesc(0x8c90688, 0xc005227868, 0xc000cb4bd0, 0x8dc4ef8, 0xc00159e360, 0x2, 0xc001966d60, 0xc000490000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/information_schema.go:2235 +0x9d
github.com/cockroachdb/cockroach/pkg/sql.postgisColumnsTablePopulator.func1(0x8c90688, 0xc005227868, 0xc000cb4bd0, 0x8dc4ef8, 0xc00159e360, 0xc0058fe800, 0xc001786708, 0xc001966e7c)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pg_extension.go:44 +0xba
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).getPlanInfo.func1.1(0x8be18e0, 0xc005738180, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:533 +0x12a
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3(0x8c90688, 0xc005227be8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x154
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc000885700, 0x8c90688, 0xc005227be8, 0xc005227a40, 0x0, 0xc0033d6fc0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x22b
```

**Expected behavior**

No crash

**Environment:**
 - CockroachDB version:  v21.2.0-alpha.00000000-4510-g8a07def384
 - Build Time: 2021/09/27 11:41:41

**Additional context**

This is from [a Django test](https://github.com/django/django/blob/4ffada36097ceccfd6f0d358217ac3c40d2a729c/tests/gis_tests/inspectapp/tests.py#L36-L56).

--- a/pkg/sql/logictest/testdata/logic_test/pg_extension
+++ b/pkg/sql/logictest/testdata/logic_test/pg_extension
@@ -1,34 +1,52 @@
 statement ok
 CREATE TABLE pg_extension_test (
   a geography(point, 4326),
-  b geometry(linestring, 3857),
+  b geometry(polygon, 3857),
   c geometry,
-  d geography
+  d geography,
+  az geography(pointz, 4326),
+  bz geometry(polygonz, 3857),
+  am geography(pointm, 4326),
+  bm geometry(polygonm, 3857),
+  azm geography(pointzm, 4326),
+  bzm geometry(polygonzm, 3857)
 )
 
 query TTTTIIT rowsort
 SELECT * FROM pg_extension.geography_columns WHERE f_table_name = 'pg_extension_test'
 ----
-test  public  pg_extension_test  a  2     4326  POINT
-test  public  pg_extension_test  d  NULL  0     GEOMETRY
+test  public  pg_extension_test  a    2     4326  POINT
+test  public  pg_extension_test  d    NULL  0     GEOMETRY
+test  public  pg_extension_test  az   3     4326  POINTZ
+test  public  pg_extension_test  am   3     4326  POINTM
+test  public  pg_extension_test  azm  4     4326  POINTZM
 
 query TTTTIIT rowsort
 SELECT * FROM pg_extension.geometry_columns WHERE f_table_name = 'pg_extension_test'
 ----
-test  public  pg_extension_test  b  2  3857  LINESTRING
-test  public  pg_extension_test  c  2  0     GEOMETRY
+test  public  pg_extension_test  b    2  3857  POLYGON
+test  public  pg_extension_test  c    2  0     GEOMETRY
+test  public  pg_extension_test  bz   3  3857  POLYGONZ
+test  public  pg_extension_test  bm   3  3857  POLYGONM
+test  public  pg_extension_test  bzm  4  3857  POLYGONZM
 
 query TTTTIIT rowsort
 SELECT * FROM geography_columns WHERE f_table_name = 'pg_extension_test'
 ----
-test  public  pg_extension_test  a  2     4326  POINT
-test  public  pg_extension_test  d  NULL  0     GEOMETRY
+test  public  pg_extension_test  a    2     4326  POINT
+test  public  pg_extension_test  d    NULL  0     GEOMETRY
+test  public  pg_extension_test  az   3     4326  POINTZ
+test  public  pg_extension_test  am   3     4326  POINTM
+test  public  pg_extension_test  azm  4     4326  POINTZM
 
 query TTTTIIT rowsort
 SELECT * FROM geometry_columns WHERE f_table_name = 'pg_extension_test'
 ----
-test  public  pg_extension_test  b  2  3857  LINESTRING
-test  public  pg_extension_test  c  2  0     GEOMETRY
+test  public  pg_extension_test  b    2  3857  POLYGON
+test  public  pg_extension_test  c    2  0     GEOMETRY
+test  public  pg_extension_test  bz   3  3857  POLYGONZ
+test  public  pg_extension_test  bm   3  3857  POLYGONM
+test  public  pg_extension_test  bzm  4  3857  POLYGONZM
 
 query ITITT
 SELECT * FROM pg_extension.spatial_ref_sys WHERE srid IN (3857, 4326) ORDER BY srid ASC

--- a/pkg/sql/pg_extension.go
+++ b/pkg/sql/pg_extension.go
@@ -64,10 +64,6 @@ func postgisColumnsTablePopulator(
 
 					var datumNDims tree.Datum
 					switch m.ShapeType {
-					case geopb.ShapeType_Point, geopb.ShapeType_LineString, geopb.ShapeType_Polygon,
-						geopb.ShapeType_MultiPoint, geopb.ShapeType_MultiLineString, geopb.ShapeType_MultiPolygon,
-						geopb.ShapeType_GeometryCollection:
-						datumNDims = tree.NewDInt(2)
 					case geopb.ShapeType_Geometry, geopb.ShapeType_Unset:
 						// For geometry_columns, the query in PostGIS COALESCES the value to 2.
 						// Otherwise, the value is NULL.
@@ -76,6 +72,17 @@ func postgisColumnsTablePopulator(
 						} else {
 							datumNDims = tree.DNull
 						}
+					default:
+						zm := m.ShapeType & (geopb.ZShapeTypeFlag | geopb.MShapeTypeFlag)
+						switch zm {
+						case geopb.ZShapeTypeFlag | geopb.MShapeTypeFlag:
+							datumNDims = tree.NewDInt(4)
+						case geopb.ZShapeTypeFlag, geopb.MShapeTypeFlag:
+							datumNDims = tree.NewDInt(3)
+						default:
+							datumNDims = tree.NewDInt(2)
+						}
+
 					}
 
 					shapeName := m.ShapeType.String()

